GolosChain/golos
# Referral program implemented #295

#295 

# Creating referral account. Hardfork check. Disabled breaking referral

1. Create `test` account and it's `testref` referral account.

```
./cli_wallet --server-rpc-endpoint=ws://127.0.0.1:8091
set_password qwer
unlock qwer
import_key 5JVFFWRLwz6JoP9kguuRFfytToGU6cLgBVTL9t6NB3D3BQLbUBS
create_account cyberfounder test "{}" "300.000 GOLOS" true
transfer_to_vesting cyberfounder test "30.000 GOLOS" true
transfer cyberfounder test "0.200 GOLOS" "" true
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.000 GOLOS"} true
```

2. You have a fail with message about unsupported operation because you done it before 19 hardfork applied. Wait for this hardfork and run again.

3. See the Mongo.

Result: account_object has an object will following fields:
```
   ...
    "referrer_account" : "test", 
    "referrer_interest_rate" : NumberLong(900), 
    "referral_end_date" : ISODate("2018-09-26T14:00:00.000+0000"), 
    "referral_break_fee_value" : 0.0, 
    "referral_break_fee_symbol" : "GOLOS", 
    ...
}
```

4. See the `get_accounts` API with following HTML page in browser:
```
<script src="node_modules/golos-js/dist/golos.min.js"></script>
Hello world
<script>
golos.config.set('websocket', 'ws://127.0.0.1:8091');
golos.api.getAccounts(['test', 'testref'], function (err, result) {
document.body.innerHTML += '<pre>' + JSON.stringify(result, null, 2) + "</pre>";
});
</script>
```
Result:
```
    "referrer_account": "test",
    "referrer_interest_rate": 900,
    "referral_end_date": "2018-09-26T14:00:00",
    "referral_break_fee": "0.000 GOLOS"
  }
]
```

5. Try to break referral program:
```
break_free_referral testref true
```
Result: fail - no right to break referral.

# Enabled breaking referral. Try to break referral without enough balance

1. Create another referral account:
```
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref2 "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.001 GOLOS"} true
```

2. Check it:
```
get_account testref2
```
Result:
```
  ...
  "referrer_account": "test",
  "referrer_interest_rate": 900,
  "referral_end_date": "2018-09-26T14:00:00",
  "referral_break_fee": "0.001 GOLOS"
}
```

3. Try to break referral (without enough balance):
```
break_free_referral testref2 true
```
Result: fail - does not have sufficient funds.

4. Transfer some money to balance and try again:
```
transfer cyberfounder testref2 "0.200 GOLOS" "" true
break_free_referral testref2 true
```
Result: success.

5. Check referral (in past) account:
```
get_account testref2
```
Result:
```
  ...
  "balance": "0.199 GOLOS",
  ...
  "referrer_account": "",
  "referrer_interest_rate": 0,
  "referral_end_date": "1970-01-01T00:00:00",
  "referral_break_fee": "0.000 GOLOS"
}
```

6. Check referrer account:
```
get_account test
```
Result:
```
  ...
  "balance": "0.001 GOLOS",
  ...
}
```

# Creating comment before breaking referral and after it

1. Create referral account:
```
create_account_referral test "0.200 GOLOS" "0.000001 GESTS" testref3 "{}" {"referrer": "test", "interest_rate": 900, "end_date": "2018-09-26T14:00:00", "break_fee": "0.001 GOLOS"} true
```

2. Post comment from it:
```
post_comment testref3 test "" test hello world "{}" true
```

3. Check comment in Mongo. Result:
```
    ...
    "beneficiaries" : [
        {
            "account" : "test", 
            "weight" : NumberLong(900)
        }
    ], 
    ...
```

4. Break referral:
```
transfer cyberfounder testref3 "0.200 GOLOS" "" true
break_free_referral testref3 true
```

5. Wait for 5 minutes and post comment again:
```
post_comment testref3 test2 "" test hello world "{}" true
```

6. Check comment in Mongo. 
Result: no object with account="test" in `beneficiaries`.
